clippy_lints/src/types.rs

```diff
@@ -307,14 +307,33 @@ fn match_type_parameter(cx: &LateContext<'_>, qpath: &QPath<'_>, path: &[&str])
     None
 }
 
+/// Checks if `qpath` has last segment with type parameter matching `diagnostic`
```

Comment:
```suggestion
/// Checks if `qpath` has last segment with type parameter matching diagnostic item `diagnostic`
```

Reply:
or call the variable diagnostic_item

Reply:
Also should this be in utils?

Reply:
Moved to utils, I had left it here because `match_type_parameter` is also in this file and I wanted to keep them together.

Reply:
Which I can remove entirely if I can figure out a way to switch this use `match_type_parameter(cx, qpath, &paths::BOX).is_some()`

But Box is a  lang item and I haven't noticed an obvious way to use a lang item there.

Reply:
match_type_parameter should also be in utils :)

Reply:
I'd be happy to move it as part of this but it is now used in a single place -- and a single place that can go away if `Box` gets a diagnostic item, or we can make a version of this function to check the lang item.

Reply:
I would still move it, and we can try to remove it as needed. It should still be in one place, otherwise it's harder to find such things.

Reply:
It looks like #6823 added these same functions in a different way, and took care of a lot of the paths I was unable. Re-basing this to just include the new diagnostic items I had added.

Will still move those 2 methods.